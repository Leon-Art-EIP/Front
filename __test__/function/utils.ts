import type { ReactTestRendererJSON } from "react-test-renderer";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import { expect } from "vitest";

function toJson(component: renderer.ReactTestRenderer): ReactTestRendererJSON {
  const result = component.toJSON();
  expect(result).toBeDefined();

  return result as ReactTestRendererJSON;
}

export function testSnapshot(element: JSX.Element): void {
  const component = renderer.create(element);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
}
