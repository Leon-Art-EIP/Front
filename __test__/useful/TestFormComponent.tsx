import { FormProvider } from "react-hook-form";
import type { ReactNode } from "react";
import useTestForm from "./useTestForm";

interface ITestInputProps {
  name: string;
  value?: string[] | string;
  children: ReactNode;
}

export default function TestFormComponent(props: ITestInputProps): JSX.Element {
  const methods = useTestForm({ [props.name]: props.value });

  return <FormProvider {...methods}>{props.children}</FormProvider>;
}
