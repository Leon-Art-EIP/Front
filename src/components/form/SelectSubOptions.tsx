import { useState, useEffect } from "react";
import { useController } from "react-hook-form";
import { IOption, IOptionSubOptions } from "../../interfaces";
import { cn } from "../../tools/cn";
import CheckIcon from "@mui/icons-material/Check";
import ExpandLessIcon from "@mui/icons-material/ExpandLess";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";

interface ISelectSubOptions {
  title?: string;
  name: string;
  options: IOptionSubOptions<string>[];
  placeholder?: string;
  className?: string;
}

export default function SelectSubOptions(props: ISelectSubOptions): JSX.Element {
  const {
    field: { onChange, value },
    fieldState: { error },
  } = useController({ name: props.name });

  const [filterCollapsed, setFilterCollapsed] = useState(false);
  const [options, setOptions] = useState<IOptionSubOptions<string>[]>(props.options);

  const toggleOptionCollapse = (optionValue: string) => {
    setOptions((prevOptions) =>
      prevOptions.map((option) =>
        option.value === optionValue ? { ...option, collapsed: !option.collapsed } : option
      )
    );
  };

  const handleSubOptionClick = (optionValue: string, subOptionValue: string) => {
    setOptions((prevOptions) =>
      prevOptions.map((option) =>
        option.value === optionValue
          ? {
              ...option,
              subOptions: option.subOptions.map((subOption) =>
                subOption.value === subOptionValue
                  ? { ...subOption, selected: !subOption.selected }
                  : subOption
              ),
            }
          : option
      )
    );
    onChange(subOptionValue);
  };

  const toggleFilterCollapse = () => {
    setFilterCollapsed(!filterCollapsed);
  };

  useEffect(() => {
    setOptions(props.options); // Update state if props.options changes
  }, [props.options]);

  return (
    <div className="flex flex-col">
      <button
        className={`flex flex-row justify-between bg-secondary items-center px-6 py-4 rounded-t ${
          !filterCollapsed && "rounded-b"
        }`}
        type="button"
        onClick={toggleFilterCollapse}
      >
        <span className="text-lg text-tertiary-hover">{props.placeholder}</span>
        {filterCollapsed ? <ExpandLessIcon /> : <ExpandMoreIcon />}
      </button>
      <div className={cn("flex flex-col gap-2 bg-secondary rounded-b", props.className)}>
        {filterCollapsed &&
          options.map((option) => (
            <div key={option.value}>
              <button
                type="button"
                className="py-2 w-full flex flex-row justify-between items-center px-4 rounded-xl duration-300 ease-in-out"
                onClick={() => toggleOptionCollapse(option.value)}
              >
                {option.label}
                {option.collapsed ? <ExpandMoreIcon /> : <ExpandLessIcon />}
              </button>
              {option.collapsed && (
                <div className="flex flex-col px-4">
                  {option.subOptions.map((subOption) => (
                    <button
                      key={subOption.value}
                      type="button"
                      className={`py-3 w-full flex flex-row justify-between items-center px-4 rounded-xl hover:bg-gray-200 hover:shadow-lg duration-300 ease-in-out ${
                        subOption.selected ? "bg-gray-200" : ""
                      }`}
                      onClick={() => handleSubOptionClick(option.value, subOption.value)}
                    >
                      {subOption.label}
                      {subOption.selected && <CheckIcon className="text-green-300" />}
                    </button>
                  ))}
                </div>
              )}
            </div>
          ))}
      </div>
      {error && <div className="text-primary">{error.message}</div>}
    </div>
  );
}
