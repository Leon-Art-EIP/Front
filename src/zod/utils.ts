export const getValueOrUndefined = (arg: unknown) =>
  arg === "" || arg === undefined || arg === null ? undefined : arg;
