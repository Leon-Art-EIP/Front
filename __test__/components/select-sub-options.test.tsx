import { test } from "vitest";
import { testSnapshot } from "../function/utils";
import TestFormComponent from "../useful/TestFormComponent";
import SelectSubOptions from "../../src/components/form/SelectSubOptions";

test("SelectSubOptions snapshot", () => {
  testSnapshot(
    <TestFormComponent name="options">
      <SelectSubOptions
        name="options"
        options={[
          {
            collapsed: false,
            label: "Option 1",
            value: "option1",
            subOptions: [
              { label: "Sub Option 1", value: "suboption1", selected: false },
              { label: "Sub Option 2", value: "suboption2", selected: false },
            ],
          },
        ]}
      />
    </TestFormComponent>
  );
});
