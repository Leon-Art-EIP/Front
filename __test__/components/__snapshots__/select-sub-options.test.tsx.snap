// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectSubOptions snapshot 1`] = `
<div
  className="flex flex-col"
>
  <button
    className="flex flex-row justify-between bg-secondary items-center px-6 py-4 rounded-t rounded-b"
    onClick={[Function]}
    type="button"
  >
    <span
      className="text-lg text-tertiary-hover"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ExpandMoreIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
    </svg>
  </button>
  <div
    className="flex flex-col gap-2 bg-secondary rounded-b"
  />
</div>
`;
